OemPkg/OemConfigPolicyCreatorPei/OemConfigPolicyCreatorPei.c

```diff
@@ -18,10 +18,68 @@
 #include <Library/ConfigVariableListLib.h>
 #include <Library/ConfigKnobShimLib.h>
 #include <Library/SafeIntLib.h>
+#include <Library/ActiveProfileIndexSelectorLib.h>
 #include <ConfigStdStructDefs.h>
 
+// these externs are provided by PlatformConfigDataLib
 extern KNOB_DATA  gKnobData[];
 extern UINTN      gNumKnobs;
+// number of profile overrides (i.e. into gProfileData)
+// this does not count the generic profile, which is not
+// in gProfileData, but rather in gKnobData's defaults
+extern UINTN    gNumProfiles;
+extern PROFILE  gProfileData[];
+
+/**
+  Helper function to apply profile overrides. This function is only
+  called if the active profile index has been validated to be within
+  the acceptable range.
+  @param[in]  ActiveProfileIndex The validated index into gProfileData
+**/
+STATIC
+VOID
+ApplyProfileOverrides (
+  UINT32  ActiveProfileIndex
+  )
+{
+  PROFILE  *ActiveProfile = &gProfileData[ActiveProfileIndex];
+  UINTN    i;
+  UINTN    Knob;
+
+  for (i = 0; i < ActiveProfile->OverrideCount; i++) {
+    Knob = ActiveProfile->Overrides[i].Knob;
+    if ((Knob >= gNumKnobs) || (Knob != gKnobData[Knob].Knob)) {
+      // something bad happened in the autogeneration, knobs are
+      // not ordered. Don't muck about finding the misordered knobs
+      // fail
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a knob autogeneration out of order, can't finish applying profile, defaulting to generic profile\n",
+        __FUNCTION__
+        ));
+
+      // we may have failed in the middle of applying the profile, so we need to
+      // write the default value for each knob we may have passed to it's cache address so we can
+      // have a clean generic profile. We didn't apply the knob for this value of i, so decrement
+      // before we start
+      i--;
+      for ( ; i > 0; i--) {
+        Knob = ActiveProfile->Overrides[i].Knob;
+        CopyMem (gKnobData[Knob].CacheValueAddress, gKnobData[Knob].DefaultValueAddress, gKnobData[Knob].ValueSize);
+      }
+
+      // do this once more when i == 0. To keep signed/unsigned comparisons issues in check, i is not signed

```

Comment:
what are the signed/unsigned issues? If you're using the same variable i, what will change?

Reply:
MSVC doesn't like a comparison between an unsigned and a signed number (and it's true, in very high number of overrides cases, you could have bad behavior). So, if I had (and I did) `i` as `INTN` so that it could go below zero and we have all the logic in the loop, MSVC spits a warning converted to an error that says you are comparing `INTN` to `UINTN` in the loop as `OverridesCount` is `UINTN`.